libsql-server/src/stats.rs

```diff
@@ -52,6 +53,100 @@ impl SlowestQuery {
     }
 }
 
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct QueryStats {
+    pub elapsed_ms: u64,
+    pub count: u64,
+    pub rows_written: u64,
+    pub rows_read: u64,
+}
+
+impl QueryStats {
+    pub fn new(elapsed_ms: u64, rows_read: u64, rows_written: u64) -> Self {
+        Self {
+            elapsed_ms,
+            count: 1,
+            rows_read,
+            rows_written,
+        }
+    }
+    pub fn empty() -> Self {
+        Self {
+            elapsed_ms: 0,
+            count: 0,
+            rows_read: 0,
+            rows_written: 0,
+        }
+    }
+    pub fn merge(&self, another: &QueryStats) -> Self {
+        Self {
+            elapsed_ms: self.elapsed_ms + another.elapsed_ms,
+            count: self.count + another.count,
+            rows_read: self.rows_read + another.rows_read,
+            rows_written: self.rows_written + another.rows_written,
+        }
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct QueriesStats {
+    #[serde(default)]
+    id: Option<Uuid>,
+
+    #[serde(default)]
+    stats_threshold: AtomicU64,
+    #[serde(default)]
+    stats: HashMap<String, QueryStats>,
+}
+
+impl QueriesStats {
+    pub fn new() -> Arc<RwLock<Self>> {
+        let mut this = QueriesStats::default();
+        this.id = Some(Uuid::new_v4());
+        Arc::new(RwLock::new(this))
+    }
+
+    pub fn register_query(&mut self, sql: &String, stat: QueryStats) {
```

Comment:
Instead of using a lock, what about using a background task and a channel. That way the cost of doing this work is offloaded from the request path and we reduce contention to just sending the query over the channel? We need to clone anyways so there is no issue there but we avoid the lock contention. It should also simplify the logic since you don't need to thread this lock everywhere?

Reply:
Another option because one challenge with a background task is that the namespace might get deallocated, is to use a concurrent hashmap that should be faster than a mutex<hashmap> https://docs.rs/dashmap/latest/dashmap/ is one option.

Reply:
As we talked on Slack, I think the background task should be used to upgrade all stats.
I've implemented an approach that does this, communicating synchronously with a 256 slots buffer.

The downside is that now we have a single thread updating stats.
If for some reason the throughput of that thread is smaller than the query throughput, it will create backpressure.

The upside is that we may consider removing/simplifying some of the internal locks.
We still need to manage the concurrency between the single reader and single writer.
But the atomic ints should be unnecessary at this point.